src/me.rs

```diff
@@ -132,61 +135,150 @@ mod native {
+    predictors.push(top);
+
+    if bo.x < fi.w_in_b - 1 {
+      let top_right = frame_mvs[(bo.y - 1) * fi.w_in_b + bo.x + 1];
+      predictors.push(top_right);
+    }
+  }
+
+  if predictors.len() > 0 {
+    let mut median_mv = MotionVector{row: 0, col: 0};
+    for mv in predictors.iter() {
+      median_mv = median_mv + *mv;
+    }
+    median_mv = median_mv / (predictors.len() as i16);
+
+    predictors.push(median_mv.quantize_to_fullpel());
+  }
+
+  predictors.push(MotionVector{row: 0, col: 0});
+
+  // Coarse motion estimation.
+
+  predictors.push(cmv.quantize_to_fullpel());
+
+  // EPZS subset C predictors.
+
+  if let Some(ref frame_ref) = frame_ref_opt {
+    let prev_frame_mvs = &frame_ref.frame_mvs[ref_slot];
+
+    if bo.x > 0 {
+      let left = prev_frame_mvs[bo.y * fi.w_in_b + bo.x - 1];
+      predictors.push(left);
+    }
+    if bo.y > 0 {
+      let top = prev_frame_mvs[(bo.y - 1) * fi.w_in_b + bo.x];
+      predictors.push(top);
+    }
+    if bo.x < fi.w_in_b - 1 {
+      let right = prev_frame_mvs[bo.y * fi.w_in_b + bo.x + 1];
+      predictors.push(right);
+    }
+    if bo.y < fi.h_in_b - 1 {
+      let bottom = prev_frame_mvs[(bo.y + 1) * fi.w_in_b + bo.x];
+      predictors.push(bottom);
+    }
+
+    predictors.push(prev_frame_mvs[bo.y * fi.w_in_b + bo.x]);
+  }
+}
+
 pub fn motion_estimation(
   fi: &FrameInvariants, fs: &FrameState, bsize: BlockSize, bo: &BlockOffset,
-  ref_frame: usize, cmv: MotionVector, pmv: &[MotionVector; 2]
+  ref_frame: usize, cmv: MotionVector, pmv: &[MotionVector; 2],
+  ref_slot: usize
 ) -> MotionVector {
   match fi.rec_buffer.frames[fi.ref_frames[ref_frame - LAST_FRAME] as usize] {
     Some(ref rec) => {
+
       let po = PlaneOffset {
         x: (bo.x as isize) << BLOCK_TO_PLANE_SHIFT,
         y: (bo.y as isize) << BLOCK_TO_PLANE_SHIFT
       };
-      let range = 16;
       let blk_w = bsize.width();
       let blk_h = bsize.height();
-      let (mvx_min, mvx_max, mvy_min, mvy_max) = get_mv_range(fi, bo, blk_w, blk_h);
-      let x_lo = po.x + ((-range + (cmv.col / 8) as isize).max(mvx_min / 8).min(mvx_max / 8));
-      let x_hi = po.x + ((range + (cmv.col / 8) as isize).max(mvx_min / 8).min(mvx_max / 8));
-      let y_lo = po.y + ((-range + (cmv.row / 8) as isize).max(mvy_min / 8).min(mvy_max / 8));
-      let y_hi = po.y + ((range + (cmv.row / 8) as isize).max(mvy_min / 8).min(mvy_max / 8));
+      let (mvx_min, mvx_max, mvy_min, mvy_max) = get_mv_range(fi.w_in_b, fi.h_in_b, bo, blk_w, blk_h);
+
+      // 0.5 is a fudge factor
+      let lambda = (fi.me_lambda * 256.0 * 0.5) as u32;
+
+      // Full-pixel motion estimation
 
       let mut lowest_cost = std::u32::MAX;
       let mut best_mv = MotionVector { row: 0, col: 0 };
 
-      // 0.5 is a fudge factor
-      let lambda = (fi.me_lambda * 256.0 * 0.5) as u32;
+      let frame_mvs = &fs.frame_mvs[ref_slot];
+      let frame_ref = &fi.rec_buffer.frames[fi.ref_frames[0] as usize];
```

Comment:
I have a doubt about this line. Does it retrieve the ReferenceFrame of the last frame?

Reply:
With our current reference structure, yes. This is probably fine for now, we don't really have a better way of mapping "frame usage" to "ref frames".

It might be interesting to try pulling candidates from all 8 references in a separate patch.